src/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs

```diff
@@ -50,37 +85,34 @@ public Task<GraphNode<RemoteResolveResult>> WalkAsync(LibraryRange library, NuGe
             }
 
             var tasks = new List<Task<GraphNode<RemoteResolveResult>>>();
-            var dependencies = node.Item.Data.Dependencies ?? Enumerable.Empty<LibraryDependency>();
+            dependencies.AddRange(node.Item.Data.Dependencies);
+
             foreach (var dependency in dependencies)
             {
-                if (predicate(dependency.Name))
+                var result = predicate(dependency.LibraryRange);
+
+                if (result == DependencyResult.Acceptable)
                 {
                     tasks.Add(CreateGraphNode(
                         dependency.LibraryRange,
                         framework,
                         runtimeName,
                         runtimeGraph,
-                        ChainPredicate(predicate, node.Item, dependency)));
-
-                    if (!string.IsNullOrEmpty(runtimeName)
-                        && runtimeGraph != null)
+                        ChainPredicate(predicate, node, dependency)));
+                }
+                else
+                {
+                    // Keep the node in the tree if we need to look at it later
```

Comment:
A further comment here explaining that we don't need to walk down this part of the graph (see my below comment) would be useful.